File: Samples for Parallel Programming with the .NET Framework/[C#][VB][C++][F#]-Samples for Parallel Programming with the .NET Framework/C# and C++ and F# and VB/C#,C++,F#,VB/ShakespeareanMonkeys/ShakespeareanMonkeys_CSharp/TextMatchGeneticAlgorithm.cs
//--------------------------------------------------------------------------
// 
//  Copyright (c) Microsoft Corporation.  All rights reserved. 
// 
//  File: TextMatchGeneticAlgorithm.cs
//
//--------------------------------------------------------------------------

using System;
using System.Linq;
using System.Text;
using System.Threading;

namespace ShakespeareanMonkeys
{
    public class TextMatchGeneticAlgorithm
    {
        private static ThreadSafeRandom _random = new ThreadSafeRandom();
        private static char[] _validChars;
        private string _targetText;
        private GeneticAlgorithmSettings _settings;
        private TextMatchGenome[] _currentPopulation;
        private bool _runParallel;

        static TextMatchGeneticAlgorithm()
        {
            // Initialize the valid characters to newlines plus all the alphanumerics and symbols
            _validChars = new char[2 + (127 - 32)];
            _validChars[0] = (char)10;
            _validChars[1] = (char)13;
            for (int i = 2, pos = 32; i < _validChars.Length; i++, pos++)
            {
                _validChars[i] = (char)pos;
            }
        }

        public TextMatchGeneticAlgorithm(bool runParallel, string targetText, GeneticAlgorithmSettings settings)
        {
            if (settings == null) throw new ArgumentNullException("settings");
            if (targetText == null) throw new ArgumentNullException("targetText");
            _runParallel = runParallel;
            _settings = settings;
            _targetText = targetText;
        }

        public void MoveNext()
        {
            // If this is the first iteration, create a random population
            if (_currentPopulation == null)
            {
                _currentPopulation = CreateRandomPopulation();
            }
            // Otherwise, iterate
            else _currentPopulation = CreateNextGeneration();
        }

        public TextMatchGenome CurrentBest { get { return _currentPopulation[0]; } }

        private TextMatchGenome[] CreateRandomPopulation()
        {
            return (from i in Enumerable.Range(0, _settings.PopulationSize)
                    select CreateRandomGenome(_random)).ToArray();
        }

        private TextMatchGenome CreateRandomGenome(Random rand)
        {
            var sb = new StringBuilder(_targetText.Length);
            for (int i = 0; i < _targetText.Length; i++)
            {
                sb.Append(_validChars[rand.Next(0, _validChars.Length)]);
            }
            return new TextMatchGenome { Text = sb.ToString(), TargetText = _targetText };
        }

        private TextMatchGenome[] CreateNextGeneration()
        {
            var maxFitness = _currentPopulation.Max(g => g.Fitness) + 1;
            var sumOfMaxMinusFitness = _currentPopulation.Sum(g => (long)(maxFitness - g.Fitness));

            if (_runParallel)
            {
                return (from i in ParallelEnumerable.Range(0, _settings.PopulationSize / 2)
                        from child in CreateChildren(
                            FindRandomHighQualityParent(sumOfMaxMinusFitness, maxFitness),
                            FindRandomHighQualityParent(sumOfMaxMinusFitness, maxFitness))
                        select child).
                        ToArray();
            }
            else
            {
                return (from i in Enumerable.Range(0, _settings.PopulationSize / 2)
                        from child in CreateChildren(
                            FindRandomHighQualityParent(sumOfMaxMinusFitness, maxFitness),
                            FindRandomHighQualityParent(sumOfMaxMinusFitness, maxFitness))
                        select child).
                        ToArray();
            }
        }

        private TextMatchGenome[] CreateChildren(
            TextMatchGenome parent1, TextMatchGenome parent2)
        {
            // Crossover parents to create two children
            TextMatchGenome child1, child2;
            if (_random.NextDouble() < _settings.CrossoverProbability)
            {
                Crossover(_random, parent1, parent2, out child1, out child2);
            }
            else
            {
                child1 = parent1;
                child2 = parent2;
            }

            // Potentially mutate one or both children
            if (_random.NextDouble() < _settings.MutationProbability) Mutate(_random, ref child1);
            if (_random.NextDouble() < _settings.MutationProbability) Mutate(_random, ref child2);

            // Return the young'ens
            return new[] { child1, child2 };
        }

        private TextMatchGenome FindRandomHighQualityParent(long sumOfMaxMinusFitness, int max)
        {
            long val = (long)(_random.NextDouble() * sumOfMaxMinusFitness);
            for (int i = 0; i < _currentPopulation.Length; i++)
            {
                int maxMinusFitness = max - _currentPopulation[i].Fitness;
                if (val < maxMinusFitness) return _currentPopulation[i];
                val -= maxMinusFitness;
            }
            throw new InvalidOperationException("Not to be, apparently.");
        }

        private void Crossover(Random rand, TextMatchGenome p1, TextMatchGenome p2, out TextMatchGenome child1, out TextMatchGenome child2)
        {
            int crossoverPoint = rand.Next(1, p1.Text.Length);
            child1 = new TextMatchGenome { Text = p1.Text.Substring(0, crossoverPoint) + p2.Text.Substring(crossoverPoint), TargetText = _targetText };
            child2 = new TextMatchGenome { Text = p2.Text.Substring(0, crossoverPoint) + p1.Text.Substring(crossoverPoint), TargetText = _targetText };
        }

        private void Mutate(Random rand, ref TextMatchGenome genome)
        {
            var sb = new StringBuilder(genome.Text);
            sb[rand.Next(0, genome.Text.Length)] = _validChars[rand.Next(0, _validChars.Length)];
            genome.Text = sb.ToString();
        }
    }

    public struct TextMatchGenome
    {
        private string _targetText;
        private string _text;

        public string Text
        {
            get { return _text; }
            set
            {
                _text = value;
                RecomputeFitness();
            }
        }

        public string TargetText
        {
            get { return _targetText; }
            set
            {
                _targetText = value;
                RecomputeFitness();
            }
        }

        private void RecomputeFitness()
        {
            if (_text != null && _targetText != null)
            {
                int diffs = 0;
                for (int i = 0; i < _targetText.Length; i++)
                {
                    if (_targetText[i] != _text[i]) diffs++;
                }
                Fitness = diffs;
            }
            else Fitness = Int32.MaxValue;
        }

        public int Fitness { get; private set; }
    }

    public class GeneticAlgorithmSettings
    {
        public int PopulationSize
        {
            get { return _populationSize; }
            set
            {
                if (value < 1 ||
                    value % 2 != 0) throw new ArgumentOutOfRangeException("PopulationSize");
                _populationSize = value;
            }
        }

        public double MutationProbability
        {
            get { return _mutationProbability; }
            set
            {
                if (value < 0 || value > 1) throw new ArgumentOutOfRangeException("MutationProbability");
                _mutationProbability = value;
            }
        }

        public double CrossoverProbability
        {
            get { return _crossoverProbability; }
            set
            {
                if (value < 0 || value > 1) throw new ArgumentOutOfRangeException("CrossoverProbability");
                _crossoverProbability = value;
            }
        }

        private int _populationSize = 30;
        private double _mutationProbability = .01;
        private double _crossoverProbability = .87;
    }
}

File: Samples for Parallel Programming with the .NET Framework/[C#][VB][C++][F#]-Samples for Parallel Programming with the .NET Framework/C# and C++ and F# and VB/C#,C++,F#,VB/ShakespeareanMonkeys/ShakespeareanMonkeys_VisualBasic/TextMatchGeneticAlgorithm.vb
'--------------------------------------------------------------------------
' 
'  Copyright (c) Microsoft Corporation.  All rights reserved. 
' 
'  File: TextMatchGeneticAlgorithm.vb
'
'--------------------------------------------------------------------------

Imports System.Text
Imports System.Threading

Namespace ShakespeareanMonkeys
	Public Class TextMatchGeneticAlgorithm
		Private Shared _random As New ThreadSafeRandom()
		Private Shared _validChars() As Char
		Private _targetText As String
		Private _settings As GeneticAlgorithmSettings
		Private _currentPopulation() As TextMatchGenome
		Private _runParallel As Boolean

		Shared Sub New()
            ' Initialize the valid characters to newlines plus all the alphanumerics and symbols.
			_validChars = New Char(2 + (127 - 32) - 1){}
			_validChars(0) = ChrW(10)
			_validChars(1) = ChrW(13)
            Dim i = 2
            Dim pos = 32
			Do While i < _validChars.Length
				_validChars(i) = ChrW(pos)
				i += 1
				pos += 1
			Loop
		End Sub

		Public Sub New(ByVal runParallel As Boolean, ByVal targetText As String, ByVal settings As GeneticAlgorithmSettings)
			If settings Is Nothing Then
				Throw New ArgumentNullException("settings")
			End If
			If targetText Is Nothing Then
				Throw New ArgumentNullException("targetText")
			End If
			_runParallel = runParallel
			_settings = settings
			_targetText = targetText
		End Sub

		Public Sub MoveNext()
            ' If this is the first iteration, create a random population.
			If _currentPopulation Is Nothing Then
				_currentPopulation = CreateRandomPopulation()
			' Otherwise, iterate
			Else
				_currentPopulation = CreateNextGeneration()
			End If
		End Sub

		Public ReadOnly Property CurrentBest() As TextMatchGenome
			Get
				Return _currentPopulation(0)
			End Get
		End Property

		Private Function CreateRandomPopulation() As TextMatchGenome()
			Return (
			    From i In Enumerable.Range(0, _settings.PopulationSize)
			    Select CreateRandomGenome(_random)).ToArray()
		End Function

		Private Function CreateRandomGenome(ByVal rand As Random) As TextMatchGenome
			Dim sb = New StringBuilder(_targetText.Length)
            For i = 0 To _targetText.Length - 1
                sb.Append(_validChars(rand.Next(0, _validChars.Length)))
            Next i
			Return New TextMatchGenome With {.Text = sb.ToString(), .TargetText = _targetText}
		End Function

		Private Function CreateNextGeneration() As TextMatchGenome()
			Dim maxFitness = _currentPopulation.Max(Function(g) g.Fitness) + 1
			Dim sumOfMaxMinusFitness = _currentPopulation.Sum(Function(g) CLng(Fix(maxFitness - g.Fitness)))

			If _runParallel Then
                Return (
                    From i In ParallelEnumerable.Range(0, _settings.PopulationSize \ 2),
                    child In CreateChildren(FindRandomHighQualityParent(sumOfMaxMinusFitness, maxFitness),
                                            FindRandomHighQualityParent(sumOfMaxMinusFitness, maxFitness))
                    Select child).ToArray()
			Else
                Return (
                    From i In Enumerable.Range(0, _settings.PopulationSize \ 2),
                    child In CreateChildren(FindRandomHighQualityParent(sumOfMaxMinusFitness, maxFitness),
                                            FindRandomHighQualityParent(sumOfMaxMinusFitness, maxFitness))
                    Select child).ToArray()
			End If
		End Function

        Private Function CreateChildren(ByVal parent1 As TextMatchGenome, ByVal parent2 As TextMatchGenome) As TextMatchGenome()
            ' Crossover parents to create two children.
            Dim child1 As TextMatchGenome = Nothing
            Dim child2 As TextMatchGenome = Nothing
            If _random.NextDouble() < _settings.CrossoverProbability Then
                Crossover(_random, parent1, parent2, child1, child2)
            Else
                child1 = parent1
                child2 = parent2
            End If

            ' Potentially mutate one or both children.
            If _random.NextDouble() < _settings.MutationProbability Then
                Mutate(_random, child1)
            End If
            If _random.NextDouble() < _settings.MutationProbability Then
                Mutate(_random, child2)
            End If

            ' Return the young'ens.
            Return {child1, child2}
        End Function

		Private Function FindRandomHighQualityParent(ByVal sumOfMaxMinusFitness As Long, ByVal max As Integer) As TextMatchGenome
            Dim val = CLng(Fix(_random.NextDouble() * sumOfMaxMinusFitness))
            For i = 0 To _currentPopulation.Length - 1
                Dim maxMinusFitness = max - _currentPopulation(i).Fitness
                If val < maxMinusFitness Then
                    Return _currentPopulation(i)
                End If
                val -= maxMinusFitness
            Next i
			Throw New InvalidOperationException("Not to be, apparently.")
		End Function

        Private Sub Crossover(ByVal rand As Random, ByVal p1 As TextMatchGenome, ByVal p2 As TextMatchGenome,
                              <System.Runtime.InteropServices.Out()> ByRef child1 As TextMatchGenome,
                              <System.Runtime.InteropServices.Out()> ByRef child2 As TextMatchGenome)
            Dim crossoverPoint = rand.Next(1, p1.Text.Length)
            child1 = New TextMatchGenome With {.Text = p1.Text.Substring(0, crossoverPoint) & p2.Text.Substring(crossoverPoint), .TargetText = _targetText}
            child2 = New TextMatchGenome With {.Text = p2.Text.Substring(0, crossoverPoint) & p1.Text.Substring(crossoverPoint), .TargetText = _targetText}
        End Sub

		Private Sub Mutate(ByVal rand As Random, ByRef genome As TextMatchGenome)
			Dim sb = New StringBuilder(genome.Text)
			sb(rand.Next(0, genome.Text.Length)) = _validChars(rand.Next(0, _validChars.Length))
			genome.Text = sb.ToString()
		End Sub
	End Class

	Public Structure TextMatchGenome
        Private _targetText As String
		Private _text As String

		Public Property Text() As String
			Get
				Return _text
			End Get
			Set(ByVal value As String)
				_text = value
				RecomputeFitness()
			End Set
		End Property

		Public Property TargetText() As String
			Get
				Return _targetText
			End Get
			Set(ByVal value As String)
				_targetText = value
				RecomputeFitness()
			End Set
		End Property

		Private Sub RecomputeFitness()
			If _text IsNot Nothing AndAlso _targetText IsNot Nothing Then
                Dim diffs = 0
                For i = 0 To _targetText.Length - 1
                    If _targetText.Chars(i) <> _text.Chars(i) Then
                        diffs += 1
                    End If
                Next i
				Fitness = diffs
			Else
				Fitness = Int32.MaxValue
			End If
		End Sub

		Private privateFitness As Integer
		Public Property Fitness() As Integer
			Get
				Return privateFitness
			End Get
			Private Set(ByVal value As Integer)
				privateFitness = value
			End Set
		End Property
	End Structure

	Public Class GeneticAlgorithmSettings
		Public Property PopulationSize() As Integer
			Get
				Return _populationSize
			End Get
			Set(ByVal value As Integer)
				If value < 1 OrElse value Mod 2 <> 0 Then
					Throw New ArgumentOutOfRangeException("PopulationSize")
				End If
				_populationSize = value
			End Set
		End Property

		Public Property MutationProbability() As Double
			Get
				Return _mutationProbability
			End Get
			Set(ByVal value As Double)
				If value < 0 OrElse value > 1 Then
					Throw New ArgumentOutOfRangeException("MutationProbability")
				End If
				_mutationProbability = value
			End Set
		End Property

		Public Property CrossoverProbability() As Double
			Get
				Return _crossoverProbability
			End Get
			Set(ByVal value As Double)
				If value < 0 OrElse value > 1 Then
					Throw New ArgumentOutOfRangeException("CrossoverProbability")
				End If
				_crossoverProbability = value
			End Set
		End Property

        Private _populationSize As Integer = 30
        Private _mutationProbability As Double = 0.01
        Private _crossoverProbability As Double = 0.87
	End Class
End Namespace